### DISCUSSION TITLE
chap3 질문 정리

### DISCUSSION THREAD

**Discussion by sihyung92** [JS-OS-study/hyokyung-and-students#5] [General]
chap3 질문 정리 스레드

**Comment by 1chz** [comment]
👍

**Comment by pullingoff** [comment]
🥳

**Comment by inseo24** [comment]
💌

**Comment by sihyung92** [comment]
# 느낀 점 & 질문
- '레지스터' 라고 하면 뭔가 하드웨어적인 것으로 느껴져서 친해지기가 어려웠는데 'CPU에서 가장 빠르게 접근 가능한 메모리 공간' 이라고 생각하니 제법 이해가 가네요!
- '컨텍스트 스위치'라는 용어에 대해 더욱 깊게 이해하게 되었습니다. 
  |before|after|
  |:--:|:--:| 
  |IO 작업을 하러 다녀오는 것도 컨텍스트 스위치겠지?|프로세스 간의 전환을 컨텍스트 스위치라고 하고, 시스템 콜을 통해 커널모드로 전환되는 것은 컨텍스트 스위치라고 부르지 않는다. 시스템 콜은 컨텍스트 스위치처럼 캐시 메모리를 비우고 새로운 메모리를 로드하는 동작을 하지는 않기 때문이다.|
  프로세스간의 컨텍스트 스위치 보다 스레드 컨텍스트 스위치가 비용이 작다던데... 이유는 몰겠당 ㅎ| 스레드 간의 컨텍스트 스위치는 동일한 프로세스 메모리를 참조할 수 있기 때문에 프로세스 컨텍스트 스위치보다 비용이 적다. |


# 문제
빈칸 채우기 + OX퀴즈 입니당 ㅎ_ㅎ

- 장기 스케줄러(Long-term scheduler, job scheduler)는 시작 프로세스 중 어떤 것들을 준비 큐로 보낼 지 결정합니다. 이는 **______** 를 제어하는 것이기 때문에 효율적인 CPU 사용에 중요한데요, time sharing system에서는 장기 스케줄러를 사용 **(합 or 하지 않습)** 니다. 
- 중기 스케줄러(Medium-Term scheduler, Swapper)는 **_______** 을 하기 위해 프로세스를 메모리에서 **________** 하는 역할을 합니다.

**Reply by 1chz** [reply]
- 장기 스케줄러(Long-term scheduler, job scheduler)는 시작 프로세스 중 어떤 것들을 준비 큐로 보낼 지 결정합니다. 이는 `degree of multiprogramming` 를 제어하는 것이기 때문에 효율적인 CPU 사용에 중요한데요, time sharing system에서는 장기 스케줄러를 사용 하지 않습니다.
- 중기 스케줄러(Medium-Term scheduler, Swapper)는 `메모리 확보`를 하기 위해 프로세스를 메모리에서 `swap out` 하는 역할을 합니다.

**Comment by inseo24** [comment]
<h3>궁금한 점</h3>
<hr/>

Q. PCB가 프로세스 관련 커널 자료구조라고 적혀 있는데 어떻게 관리될까?

A. Linked List 방식으로 관리가 된다. PCB List Head에 PCB들이 생성될 때마다 붙게 되는데 주소값으로 연결이 된 연결 리스트 형태로 삽입, 삭제가 용이하다. 프로세스가 생성되면 해당 PCB가 생성되고 프로세스 완료 시 제거된다.
(수행중인 프로세스를 변경할 때, CPU의 레지스터 정보가 변경되는 것이 Context switch)

Q. Context switch에 발생하는 다른 cost에는 뭐가 있을까?

1. cache memory flush(cache 초기화)
2. Memory mapping 초기화
3. 더 있다면 뭐가 있을까요? ← 모르겠음


<h3>문제</h3>
<hr/>


Q. context switching에 대해서 설명하세요.

**Reply by 1chz** [reply]
프로세스의 상태들을 변경하고 CPU를 재할당하는 것

**Reply by sihyung92** [reply]
CPU에서 실행되는 프로세스를 다른 프로세스로 전환하는 것을 컨텍스트 스위칭이라고 부른다

**Reply by 1chz** [reply]
컨텍스트 스위칭으로 인해 `Timer`가 할당한 시간을 까먹는 것 ...?

**Reply by pullingoff** [reply]
CPU를 넘겨줄때(재할당할 때) 넘겨주는 프로세스가 PCB에 문맥을 저장하고, 넘겨받는 프로세스는 자기 문맥을 하드웨어로 복원하는 것

**Comment by 1chz** [comment]
# 느낀점

---

- `프로세스`와 `스레드`에 대해 조금 더 알게 됨으로써 `Java`에 대해 조금 더 이해할 수 있게 됐다.
- JDK를 분석하다 보면 `volatile`가  많이 보였는데, 이 키워드를 사용하는 이유에 대해 이해하게 됐다.

# 문제

---

- 운영체제가 스케쥴링을 통해 CPU를 할당하는 대상은 무엇인가?

**Reply by sihyung92** [reply]
프로세스~!

**Reply by inseo24** [reply]
프로세스~!!

**Reply by sihyung92** [reply]
정답 : 커널 스레드

**Reply by pullingoff** [reply]
운영체제 안의 커널 프로세스가 할당 받아서 사용자 프로세스를 제어한다 는거죠!!!???

**Reply by 1chz** [reply]
> 운영체제 안의 커널 프로세스가 할당 받아서 사용자 프로세스를 제어한다 는거죠!!!???

운영체제안의 스레드(커널 스레드)가 CPU를 할당받고 커널 스레드가 LWP를 통해 유저 스레드와 일대일로 매핑된다네요 😊

**Comment by pullingoff** [comment]
# 느낀점
- 이번 장을 통해 컴퓨터가 어떻게 한번에 많은 일을 하는가 알게 되었다. CPU 스케줄러에 종류가 있는지 몰랐는데 각각의 역할에 대해 알게 되었다.

# 질문
각자 멀티 프로세싱이나 멀티 스레드를 사용해본 적이 있는지, 있다면 그 경험에 대해 공유해주세요! 

# 문제
1. PCB의 구성요소 3개를 말하시오.
2. 인터럽트가 발생할 때마다 문맥 교환이 일어난다 (O,X)?
3. 왜 현대의 시분할 시스템엔 보통 장기 스케줄러가 없을까?

**Reply by sihyung92** [reply]
프로그램 카운터, CPU 레지스터, 프로세스의 메모리 위치정보(코드, 스택, 데이터)

**Reply by inseo24** [reply]
2. x
3. 장기 스케줄러는 어떤 프로세스를 ready queue에 넣어주는 역할을 하는데, 현대의 운영체제에는 프로세스가 시작되면 바로 그 프로세스에 메모리를 할당해 ready queue에 넣어줘서 필요가 없음.

**Reply by pullingoff** [reply]
1. PID
2. 프로세스 상태
3. 프로그램 카운터 (프로세스가 다음에 실행해야할 명령어가 저장된 주소)
4. 프로세스의 소유자 정보
5. cpu 스케줄링 관련 정보 (우선 순위 등)
6. 마지막으로 실행된 일시
7. cpu 레지스터 정보
8. 부모 프로세스와 자식 프로세스로의 포인터, 할당된 자원에 대한 포인터 정보
